pkg/kubectl/apply/parse/factory.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parse
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/kubectl/apply"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/util/openapi"
+)
+
+// Factory creates an Element by combining object values from recorded, local and remote sources with
+// the metadata from an openapi schema.
+type Factory struct {
+	// Resources contains the openapi field metadata for the object models
+	Resources openapi.Resources
+}
+
+// CreateElement returns an Element by collating the recorded, local and remote field values
+func (b *Factory) CreateElement(recorded, local, remote map[string]interface{}) (apply.Element, error) {
+	// Create an Item from the 3 values.  Use empty name for field
+	visitor := &ElementBuildingVisitor{b.Resources}
+
+	gvk, err := getCommonGroupVersionKind(recorded, local, remote)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the openapi object metadata
+	s := visitor.resources.LookupResource(gvk)
+	oapiKind, err := getKind(s)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the item for the type
+	hasRecorded := recorded != nil
+	hasLocal := local != nil
+	hasRemote := remote != nil
+	fieldName := ""
+	item, err := visitor.getItem(oapiKind, fieldName, recorded, local, remote, hasRecorded, hasLocal, hasRemote)
```

Comment:
8 parameters doesn't looks so good, maybe put `recorded, local, remote, hasRecorded, hasLocal, hasRemote` into a struct.
And `primitiveItem`, `listItem`, `mapItem` and `typeItem` in `pkg/kubectl/apply/parse/item.go` can also benefit from it by embedding it. Reduce a lot of code dup.

Reply:
I think that's fine. The goal of this function is to build the structure anyway?

Reply:
It is a lot of arguments and obfuscates their purpose.  I pulled them into variables so they are self documenting.  I see you point about including them in a common structure.  Will address that.